state/state.go

```diff
@@ -1118,6 +1118,47 @@ var (
 	errLocalApplicationExists          = errors.Errorf("application already exists")
 )
 
+type AddCloudServiceArgs struct {
+	// Id will be the application Name if it's a part of application,
+	// and will be controller UUID for k8s a controller(controller does Not have an application).
+	Id         string
+	ProviderId string
+	Addresses  []network.Address
+}
+
+// AddCloudService creates a cloud service.
+func (st *State) AddCloudService(args AddCloudServiceArgs) (_ *CloudService, err error) {
+	defer errors.DeferredAnnotatef(&err, "cannot add cloud service %q", args.ProviderId)
+
+	doc := cloudServiceDoc{
+		DocID:      args.Id,
+		ProviderId: args.ProviderId,
+		Addresses:  fromNetworkAddresses(args.Addresses, OriginProvider),
+	}
+	svc := newCloudService(st, &doc)
+
+	ops, err := svc.saveServiceOps(doc)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err = st.db().RunTransaction(ops); err == nil {
+		if err = svc.Refresh(); err != nil {
```

Comment:
This refresh after save pattern is not one we've used elsewhere have we?
Would prefer to stick to our standard save patterns

Reply:
we should be using st.db().Run(buildTxn) if we can, instead of RunTransaction, as it allows you to revalidate your assertions and give better error messages along with retrying if the db has changed vs memory.